### PR TITLE
Training many singletask networks and Preliminary Gradient Support

### PR DESCRIPTION
**Pull request by rbharath**
More details once tests pass.

### CODE REVIEW

**Comment by rbharath** [comment]
Most of this PR works on merging in code by @joegomes into deepchem.

**Comment by rbharath** [comment]
This PR includes changes to:

- Train Tensorflow models with `SingletaskToMultitask` where there are hundreds of tasks. Previously, the `store_in_memory` option made it so that all allocated Tensorflow memory stayed allocated (this is apparently due to a bug where `session.close()` doesn't release memory properly if the `session` object still lives). The solution basically makes `store_in_memory` obsolete. Tested with ~200 tasks.
- Adds `AtomicCoordinates` featurizer by @joegomes. Featurizers a molecule as a `(N,3)` tensor, where `N` is the number of atoms. 
- Adds preliminary support for working with models that predict physical forces on system. Again, work by @joegomes. The current code is functional, but quite messy. Some refactoring, perhaps the introduction of a `PhysicalModel` class is required to clean this situation up.

@joegomes: Could you please review the parts of this PR that have to do with the physical force support? I want to make sure I didn't make any unreasonable changes or leave important code out.

File: deepchem/datasets/__init__.py [hunk 1]
```diff
@@ -682,6 +682,40 @@ def _update_mean_and_std(self, df):
...
       save_to_disk(yss, os.path.join(self.data_dir, row['y_sum_squares']))
 
+  def get_grad_statistics(self):
+    """Computes and returns statistics of this dataset
+
+    This function assumes that the first task of a dataset holds the energy for
+    an input system, and that the remaining tasks holds the gradient for the
+    system.
+
+    TODO(rbharath, joegomes): It is unclear whether this should be a Dataset
```
Comment (joegomes) [inline deepchem/datasets/__init__.py]: I agree, I initially added this here in the spirit of keeping everything that access the metadata_df in the dataset object but currently would only be used by PES fitting

File: deepchem/featurizers/atomic_coordinates.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    Parameters
+    ----------
+    mol : RDKit Mol
+          Molecule.
+    """
+
+    N = mol.GetNumAtoms()
+    coords = np.zeros((N,3))
+    # TODO(joegomes, rbharath): Add comment about magic number
```
Comment (joegomes) [inline deepchem/featurizers/atomic_coordinates.py]: comment would be something like
"RDKit stores atomic coordinates in Angstrom.  Atomic unit of length is the bohr (1 bohr = 0.529177 Angstrom).  Converting units makes gradient calculation consistent with most QM software packages."

File: deepchem/metrics/__init__.py [hunk 3]
```diff
@@ -181,7 +182,12 @@ def compute_metric(self, y_true, y_pred, w=None, n_classes=2, filter_nans=True):
       if filter_nans:
         computed_metrics = np.array(computed_metrics)
         computed_metrics = computed_metrics[~np.isnan(computed_metrics)]
-      return self.task_averager(computed_metrics)
+      if self.compute_force_metrics:
```
Comment (joegomes) [inline deepchem/metrics/__init__.py]: This should be something like "compute_energy_metric" since we are printing out the force error but only passing back the energy error metric.  This is used where we fit the energy and gradient with multitask models but averaging the metrics of force and energy didn't make much sense

File: deepchem/models/__init__.py [hunk 4]
```diff
@@ -148,8 +154,195 @@ def predict(self, dataset, transformers=[]):
...
+    # Special case to handle singletasks.
+    if n_tasks == 1:
+      y_pred = np.reshape(y_pred, (n_samples,)) 
     return y_pred
 
+  def predict_grad(self, dataset, transformers=[]):
+    """
+    Uses self to calculate gradient on provided Dataset object.
+
+    TODO(rbharath): Should we assume each model has meaningful gradients to
```
Comment (joegomes) [inline deepchem/models/__init__.py]: @rbharath This is added here because `HyperparamOpt` class returns the `Model` object as `best_model` rather than the `TensorFlowModel` object.  I think ideally this would be in the `TensorFlowAtomicRegressor` family of classes but due to how I was optimizing models and testing gradients I had to put it here.

File: deepchem/models/__init__.py [hunk 5]
```diff
@@ -148,8 +154,195 @@ def predict(self, dataset, transformers=[]):
...
+    
+    return energy_error, grad_error
+
+  def evaluate_error_class2(self, dataset, transformers=[]):
+    """
+    Evaluate the error in energy and gradient components, forcebalance-style.
+
+    TODO(rbharath): Should be a subclass PhysicalModel method. Also, need to
+    find a better name for this method (class2 doesn't tell us anything about the
+    semantics of this method.
```
Comment (joegomes) [inline deepchem/models/__init__.py]: @rbharath I agree. Along the lines of my above comment, this should be the compute error function of the `AtomicRegressor` class that trains only on energy but predicts forces with tf.gradients.

File: deepchem/models/__init__.py [hunk 6]
```diff
@@ -148,8 +154,195 @@ def predict(self, dataset, transformers=[]):
...
+    print("Grad error (RMSD): %f kJ/mol/A" % grad_error)
+    
+    return energy_error, grad_error
+
+  def test_fd_grad(self, dataset, transformers=[]):
+    """
+    Uses self to calculate finite difference gradient on provided Dataset object.
+    Currently only useful if your task is energy and self contains predict_grad_on_batch.
+
+    TODO(rbharath): This shouldn't be a method of the Model class. Perhaps a
```
Comment (joegomes) [inline deepchem/models/__init__.py]: @rbharath This could also be put into `AtomicRegressor` classes if we re-factor hyperparamopt

File: deepchem/transformers/__init__.py [hunk 7]
```diff
@@ -95,7 +100,7 @@ def _transform_row(i, df, transformer, data_dir):
 class NormalizationTransformer(Transformer):
 
   def __init__(self, transform_X=False, transform_y=False, transform_w=False,
-               dataset=None):
+               dataset=None, transform_gradients=False):
```
Comment (joegomes) [inline deepchem/transformers/__init__.py]: @rbharath Standard Normalization transformer is used for option 1 and 3 where we are not training on tf.gradient(E).  `untransform_grad` is used in the predict grad function of option 3.  I had two different version of NormalizationTransform because this one was fine for options 1 and 3, but option two needed the gradient to be properly transformed during model.fit()

**Comment by joegomes** [comment]
@rbharath Looks like it all got in to this PR.  I've reviewed the code and left comments on open questions.  Thanks!

File: deepchem/datasets/__init__.py (hunk 1 above)
Comment (rbharath) [inline deepchem/datasets/__init__.py]: I think the main issue preventing this from being a user-space function is the use of `y_n`. I think this can be avoided by just using `y.shape[0]` directly. Not going to refactor this in the current PR, but have added a note so one of us can clean it up in a follow-on.

File: deepchem/featurizers/atomic_coordinates.py (hunk 2 above)
Comment (rbharath) [inline deepchem/featurizers/atomic_coordinates.py]: Added!

File: deepchem/metrics/__init__.py (hunk 3 above)
Comment (rbharath) [inline deepchem/metrics/__init__.py]: Done. I think this should be moved into user space by making it a `task_averager` function. Not to be done in this PR, but I've added a line note as a comment.

File: deepchem/models/__init__.py (hunk 5 above)
Comment (rbharath) [inline deepchem/models/__init__.py]: Let's talk more about the right refactoring strategy IRL. For now, I think it's harmless to leave these methods in here, but let's pick a coherent strategy for moving them out in the next few PRs.

File: deepchem/transformers/__init__.py (hunk 7 above)
Comment (rbharath) [inline deepchem/transformers/__init__.py]: @joegomes I think the general question is whether this is a general transformation that we expect to be useful elsewhere. I do think we expect to have many models that work with gradients, but what's the right abstraction to work with these models? Let's chat IRL about these design questions. I think it's harmless to leave this code in here for now.

**Comment by rbharath** [comment]
@joegomes Added in feedback, and responded to in-line comments. I think I'm going to go ahead and merge this one in once tests pass. Let's refactor the edges out in follow on PRs.
